src/scheduler/Executor.cpp

```diff
@@ -10,45 +10,60 @@
 #include <faabric/util/timing.h>
 
 namespace faabric::scheduler {
+
+// TODO - avoid the copy of the message here?
 Executor::Executor(const faabric::Message& msg)
   : boundMessage(msg)
+  , threadPoolSize(faabric::util::getUsableCores())
+  , threadPoolThreads(threadPoolSize)
+  , threadQueues(threadPoolSize)
 {
     faabric::util::SystemConfig& conf = faabric::util::getSystemConfig();
 
-    threadPoolSize = faabric::util::getUsableCores();
+    assert(!boundMessage.user().empty());
+    assert(!boundMessage.function().empty());
 
     // Set an ID for this Executor
     id = conf.endpointHost + "_" + std::to_string(faabric::util::generateGid());
     faabric::util::getLogger()->debug("Starting executor {}", id);
 }
 
-Executor::~Executor()
-{
-    finish();
-}
+Executor::~Executor() {}
 
 void Executor::finish()
 {
-    // Shut down thread pool with a series of kill messages
-    for (auto& queuePair : threadQueues) {
-        std::shared_ptr<BatchExecuteRequest> killReq =
-          faabric::util::batchExecFactory();
+    const auto& logger = faabric::util::getLogger();
+    logger->debug("Executor {} shutting down", id);
 
-        killReq->add_messages()->set_type(faabric::Message::KILL);
-        queuePair.second.enqueue(std::make_pair(0, killReq));
-    }
+    // Shut down thread pools and wait
+    for (int i = 0; i < threadPoolThreads.size(); i++) {
+        if (threadPoolThreads.at(i) == nullptr) {
+            continue;
+        }
```

Comment:
This bit of code was previously adding a message to every queue regardless of whether there was a thread waiting at the end of it.